woocommerce/pinterest-for-woocommerce
# Feed Refactor with Action Scheduler Framework.

This supersedes the #271. There was a bad merge with `develop`. I wanted this branch to start with a new develop merge - this time not botched. 


### Changes proposed in this Pull Request:
This PR updates the feed generation process.
The main goals for the PR:

- simplified and streamlined generation process
- better error handling
- multi-feed support on the server-side
- predictable scalability
- extensibility of the process

Bonus:
- removes variable parents from XML ( bug #242 )
- if https://github.com/woocommerce/pinterest-for-woocommerce/pull/284 will be merged then additionally #206 will be fixed.

There are very small changes in the UI compared to the original implementation. The feed status no longer shows the total number of products for processing. The rationale for this change comes from two sources:
1. We no longer precalculate the products list and store it in a DB option. Instead, the implementation fetches pages of product ids during each generation cycle. The presence of a large DB option containing hundreds of thousands of items that would be read and updated during batches generation was flagged as a problem by merchants that have used Facebook For WooCommerce plugin. This plugin was using the same method. After changing to the pagination method the merchants have observed better performance. Learning from that experience I have used the same approach here.
2. When we will have multiple feeds the total number of products may not be a very useful number bc each feed may have a different total number.

Usage of Action Scheduler Jobs Framework ( https://github.com/woocommerce/action-scheduler-job-framework ) will be highly beneficial when we will introduce multiple feed generation. During each generation cycle, we will be able to determine if a particular product should be a part of a particular feed and the `ProductsXmlFeed` class will understand the location parameter passed ( currently it does nothing ) to determine the shape of the feed element for that feed.

Most of the parts of the code are ready for multiple feeds. During the refactor it has become obvious that if this capability would not be implemented alongside the refactoring it would be a wasted effort. There are places that still assume that we have a singular feed. Those places need to wait for UI changes part of the Multiple-feeds feature.

This is based on improvements from https://github.com/woocommerce/pinterest-for-woocommerce/pull/240 ( now in develop )

 =


Closes #257
Closes #242
Closes #185 

### Screenshots:




### Detailed test instructions:
Please reinstall composer dependencies - this uses a new library and if not installed this will fail ( easiest way would be to just delete vendors folder and then composer install )

There are two distinct scenarios:

- First: fresh install
1. Install the plugin.
2. Make the connection.
3. Everything should work as usual.

- Second: updating over an existing version
0. Install from th developer branch, make the connection.
1. Update the plugin, remember to reinstall composer deps.
2. Old actions should be removed
3. Everything should work as usual except that the feed generation is now controlled by the Action Scheduled framework.


### Changelog entry
Fix - Remove parent variable products from the feed file. ( #242 )
Update - Feed generation process refactor. ( #185 )
Add - Multiple feed backend.
Tweak - Additional logging of the feed generation process.
>
